src/notification/index.js

```diff
@@ -211,16 +211,41 @@ export class NotificationListener {
 
   /** Private. */
   handleDeviceToken = async (deviceToken: mixed) => {
-    // A device token should normally be a string of hex digits. Sometimes,
-    // however, we appear to receive objects here. Log this. (See issue #3672.)
-    if (typeof deviceToken !== 'string' || deviceToken === '[Object object]') {
-      // $FlowFixMe: deviceToken probably _is_ JSONable, but we can only hope
-      const token: JSONable = deviceToken;
-      logging.error('Received invalid device token', { token });
+    /** Flow-comprehensible validator for `deviceToken`. */
+    const validateToken = (): string | null | void => {
+      // Null device tokens are known to occur (at least) on Android emulators
+      // without Google Play services, and have been reported in other
+      // scenarios.
+      //
+      // See https://stackoverflow.com/q/37517860 for relevant discussion.
+      if (deviceToken === null) {
+        return deviceToken;
+      }
+
+      // A non-null device token should normally be a string of hex digits.
+      // Sometimes, however, we appear to receive objects here. Log this. (See
+      // issue #3672.)
+      //
+      // (The '[Object object]' case is not expected, but it's possible we're
+      // receiving the object as an already-improperly-flattened string.)
+      if (typeof deviceToken !== 'string' || deviceToken === '[Object object]') {
+        // $FlowFixMe: deviceToken probably _is_ JSONable, but we can only hope
+        const token: JSONable = deviceToken;
+        logging.error('Received invalid device token', { token });
+        // Take no further action.
+        return undefined;
+      }
+
+      // If we reach this point, `deviceToken` must be a string.
+      return (deviceToken: string);
+    };
+
+    const token: ?string = validateToken();
+    if (token === undefined) {
       return;
     }
 
```

Comment:
I'm afraid I find this way of writing the logic really twisty. :-/ This code is significant work for me to follow and be confident I understand; the logic it's ultimately implementing isn't complicated, so it doesn't seem like that should have to be the case.

One thing I think is contributing to that effect is the use of a closure, where this helper function takes `deviceToken` implicitly from an enclosing scope rather than explicitly as a parameter. Another is the sort of alternation in the control flow: first there's a path that accepts, then one that rejects, then one that accepts... and that combines with the accept/reject distinction being expressed in a non-obvious and implicit way.

How about we try some edits to this code while both looking at the same screen together sometime in the next day or two? That might be an effective way to discuss.